urleb43757003_Origraena_mas_platform_tgz-pJ8-ori_MainJ8/src/ori/mas/core/DefaultHeart.java

```diff
@@ -0,0 +1,284 @@
+	 * </p>
+	 * @param i The influence to apply.
+	 * @return The influence which has really been applied (may be <code>null</code>).
+	 */
+	protected Influence makeInfluence(Influence i) {
+		if (i instanceof MultipleInfluence) {
+			for (Influence j : (MultipleInfluence)i)
+				submitPrioritaryInfluence(j);
+			return i;
+		}
+		if (i instanceof PhysicalInfluence)
+			return this.makeInfluence((PhysicalInfluence)i);
+		if (i instanceof BirthInfluence)
+			return this.makeInfluence((BirthInfluence)i);
+		return null;
+	}
+
+	protected Influence makeInfluence(PhysicalInfluence i) {
+		if (i.target() == null)
+			return null;
+		if (i instanceof MovementInfluence)
+			return this.makeInfluence((MovementInfluence)i);
+		if (i instanceof CollideInfluence)
+			return this.makeInfluence((CollideInfluence)i);
+		if (i instanceof ChangeBodyPropertyInfluence)
+			return this.makeInfluence((ChangeBodyPropertyInfluence)i);
+		if (i instanceof DeathInfluence)
+			return this.makeInfluence((DeathInfluence)i);
+		return null;
+	}
+
+	protected Influence makeInfluence(MovementInfluence i) {
+		Body target = i.target();
+		if (target == null)
+			return null;
+		Point vector = i.vector();
+		if (vector == null)
+			return null;
+		List<Point> tmp;
+		// TODO set comparator
+		Segment wayLine;
+		int minX = Integer.MAX_VALUE;
+		int minY = Integer.MAX_VALUE;
+		int maxX = Integer.MIN_VALUE;
+		int maxY = Integer.MIN_VALUE;
+		for (Point p : target.shape().boundingShape().points()) {
+			if (p.x < minX)
+				minX = p.x;
+			if (p.y < minY)
+				minY = p.y;
+			if (p.x > maxX)
+				maxX = p.x;
+			if (p.y > maxY)
+				maxY = p.y;
+			if (p.x + vector.x < minX)
+				minX = p.x + vector.x;
+			if (p.y + vector.y < minY)
+				minY = p.y + vector.y;
+			if (p.x + vector.x > maxX)
+				maxX = p.x + vector.x;
+			if (p.y + vector.y > maxY)
+				maxY = p.y + vector.y;
+		}
+		Point min,max;
+		min = new Point(minX,minY);
+		max = new Point(maxX,maxY);
+
+		Surface<Body> s = _world.scene().getPartlyIn(new Rectangle(min,max));
+		Body collidingBody = null;
+		Point collidingPoint = null;
+		int d = Integer.MAX_VALUE;
+
+		for (Point p : target.shape().boundingShape().points()) {
+			wayLine = new Segment(p.x,p.y,p.x+vector.x,p.y+vector.y);
+			for (Body b : s) {
+				if ((b != target) && (this.mayCollide(target,b))) {
+					tmp = Shapes.collidePoints(wayLine,b);
+					for (Point p2 : tmp) {
+						if (p2.euclidian(p) < d) {
+							d = p2.euclidian(p);
+							collidingBody = b;
+							collidingPoint = p2;
+							vector.x = p2.x;
+							vector.y = p2.y;
+							vector.minus(p);
+						}
+					}
+				}
+			}
+		}
+		
+		this.callbackMove(i);
+		if (collidingBody != null)
+			this.submitPrioritaryInfluence(new CollideInfluence(target,collidingBody,collidingPoint));
+		return i;
+	}
+
+	protected Influence makeInfluence(CollideInfluence i) {
+		//System.out.println("colliding : "+i);
+		return null;
+	}
+
+	protected Influence makeInfluence(ChangeBodyPropertyInfluence i) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.